Add support for visualizing bones in Advanced Import Settings

jeronimo-schreyer opened the pull request:
When an imported model Skeleton3D type node is selected, the bones are drawn using lines or octahedrons to provide a clearer reference to their position. Refactored Skeleton3DGizmoPlugin::redraw now uses a static function to generate bone meshes

Update
editor/import/scene_import_settings.cpp
editor/import/scene_import_settings.h
editor/plugins/skeleton_3d_editor_plugin.h
editor/plugins/skeleton_3d_editor_plugin.cpp

- *Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/6801.*
- this PR comes from https://github.com/godotengine/godot/pull/76818

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

Calinou commented on editor/import/scene_import_settings.cpp:
```diff
@@ -1404,6 +1412,14 @@ SceneImportSettings::SceneImportSettings() {
+		bones_mesh_preview = memnew(MeshInstance3D);
+		bones_mesh_preview->set_cast_shadows_setting(GeometryInstance3D::SHADOW_CASTING_SETTING_OFF);
+		bones_mesh_preview->set_skeleton_path(NodePath(""));
+		bones_mesh_preview->translate(Vector3(0.0, 0.0, 0.035));
```
Out of curiosity, why is this `translate()` call needed?

Calinou replied:
@jeronimo-schreyer Bump :slightly_smiling_face:

fire commented:
Can you merge into one commit?

SaracenOne commented on editor/import/scene_import_settings.cpp:
```diff
@@ -1598,6 +1606,14 @@ SceneImportSettings::SceneImportSettings() {
+		bones_mesh_preview = memnew(MeshInstance3D);
+		bones_mesh_preview->set_cast_shadows_setting(GeometryInstance3D::SHADOW_CASTING_SETTING_OFF);
+		bones_mesh_preview->set_skeleton_path(NodePath(""));
+		bones_mesh_preview->translate(Vector3(0.0, 0.0, 0.035));
```
This seems to be a mistake.
![godot windows editor dev x86_64_n6ulT8u9mz](https://github.com/godotengine/godot/assets/12756047/26d255fa-e392-4c6f-b553-eae8046c7897)

SaracenOne reviewed (changes requested):
This will probably be fine once the translation thing is addressed and the PR is rebased.

jeronimo-schreyer committed:
add support for visualizing bones in Advanced Import Settings

When an imported model Skeleton3D type node is selected, the bones are drawn using lines or octahedrons to provide a clearer reference to their position.
Refactored Skeleton3DGizmoPlugin::redraw now uses a static function to generate bone meshes

SaracenOne reviewed (approved):
LGTM👍

YuriSizov commented:
Could you please amend your commit message to remove this part?

```
Update
editor/import/scene_import_settings.cpp
editor/import/scene_import_settings.h
editor/plugins/skeleton_3d_editor_plugin.h
editor/plugins/skeleton_3d_editor_plugin.cpp
Conflicts:
	editor/import/scene_import_settings.cpp
	editor/import/scene_import_settings.h
```

It's not relevant to the changes and doesn't need to be stored in the git history.

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -1188,16 +1188,17 @@ int Skeleton3DEditor::get_selected_bone() const {
 }
 
+Skeleton3DGizmoPlugin::SelectionMaterials Skeleton3DGizmoPlugin::selection_materials;
 Skeleton3DGizmoPlugin::Skeleton3DGizmoPlugin() {
```
```suggestion

Skeleton3DGizmoPlugin::Skeleton3DGizmoPlugin() {
```

YuriSizov commented on editor/import/scene_import_settings.cpp:
```diff
@@ -411,6 +412,11 @@ void SceneImportSettings::_fill_scene(Node *p_node, TreeItem *p_parent_item) {
 		}
 	}
+
+	Skeleton3D *skeleton = Object::cast_to<Skeleton3D>(p_node);
+	if (skeleton) {
+		bones_mesh_preview->set_mesh(Skeleton3DGizmoPlugin::get_bones_mesh(skeleton, -1, true));
+	}
```
This relationship doesn't seem right to me. We shouldn't rely directly on plugin code here, that defeats the purpose of it being a plugin. We also can't use the same pattern going forward if we need to bring more viewport gizmos to the advanced import dialog preview.

I think it's time to integrate viewport/gizmo plugins into this preview, or alternatively devise a separate pluggable interface.

lyuma replied:
We talked about this at the animation asset meeting and we would like this change. Saracen has already approved the technical aspects.

I don't think this is such a big issue since it's a static function and both are in editor/ so there is no linking problem.
@AThousandShips would it be possible to review this usage of a plugin function from the import code and determine if this is ok?

AThousandShips replied:
I'll take a look :)

AThousandShips replied:
I don't think there's any issue with this, both are fully contained in the editor, and it saves a lot of code duplication, could theoretically move it to Skeleton3D or similar but it being a plugin here doesn't change anything in my opinion, as it's not a module or similar but fully integrated

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.h:
```diff
@@ -253,11 +253,15 @@ class Skeleton3DEditorPlugin : public EditorPlugin {
+	static SelectionMaterials selection_materials;
 
 public:
+	static Ref<ArrayMesh> get_bones_mesh(Skeleton3D *skeleton, int selected, bool is_selected);
```
Please apply this and my other suggestions

AThousandShips commented on editor/import/scene_import_settings.cpp:
```diff
@@ -1607,6 +1615,13 @@ SceneImportSettings::SceneImportSettings() {
+	{
+		bones_mesh_preview = memnew(MeshInstance3D);
+		bones_mesh_preview->set_cast_shadows_setting(GeometryInstance3D::SHADOW_CASTING_SETTING_OFF);
+		bones_mesh_preview->set_skeleton_path(NodePath(""));
```
```suggestion
		bones_mesh_preview->set_skeleton_path(NodePath());
```

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -1226,6 +1228,11 @@ void fragment() {
 }
 
+Skeleton3DGizmoPlugin::~Skeleton3DGizmoPlugin() {
+	selection_materials.unselected_mat = Ref<StandardMaterial3D>();
+	selection_materials.selected_mat = Ref<ShaderMaterial>();
```
```suggestion
	selection_materials.unselected_mat.unref();
	selection_materials.selected_mat.unref();
```

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -1188,16 +1188,18 @@ int Skeleton3DEditor::get_selected_bone() const {
-
-	selected_mat = Ref<ShaderMaterial>(memnew(ShaderMaterial));
-	selected_sh = Ref<Shader>(memnew(Shader));
+	selection_materials.unselected_mat = Ref<StandardMaterial3D>(memnew(StandardMaterial3D));
```
```suggestion
	selection_materials.unselected_mat.instantiate();
```
Cleaner IMO

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -1188,16 +1188,18 @@ int Skeleton3DEditor::get_selected_bone() const {
+	selection_materials.unselected_mat->set_flag(StandardMaterial3D::FLAG_SRGB_VERTEX_COLOR, true);
+	selection_materials.unselected_mat->set_flag(StandardMaterial3D::FLAG_DISABLE_FOG, true);
+
+	selection_materials.selected_mat = Ref<ShaderMaterial>(memnew(ShaderMaterial));
```
```suggestion
```

AThousandShips commented on editor/plugins/skeleton_3d_editor_plugin.cpp:
```diff
@@ -1216,7 +1218,7 @@ void fragment() {
 }
 )");
-	selected_mat->set_shader(selected_sh);
+	selection_materials.selected_mat->set_shader(selected_sh);
```
```suggestion
	selection_materials.selected_mat.instantiate();
	selection_materials.selected_mat->set_shader(selected_sh);
```
Keep them local IMO

lyuma commented:
@jeronimo-schreyer do you think you will have time to resolve the pending review feedback?
If not, I can try take a look at it later this week

SaracenOne reviewed (changes requested):
I feel like the skeleton should match the animations when they are being played back. It currently doesn't seem to do that when I tested it.

clayjohn commented:
> @jeronimo-schreyer do you think you will have time to resolve the pending review feedback? If not, I can try take a look at it later this week

Maybe you should just take over. It would be really nice to have this done for 4.3 and we don't have much more time

akien-mga commented:
Superseded by #96094.
